### DISCUSSION TITLE
Fastify is very slow when I added many routes.

### DISCUSSION THREAD

**Discussion by meotimdihia** [fastify/fastify#3476] [Q&A] [answered]
Currently, it takes 7s for me to start the Fastify server, and with the VScode debugger attached, it would take 10s+.
Do we have a way to optimize this problem?

https://user-images.githubusercontent.com/300961/142988192-440775d0-c2d5-4629-9fcf-5c094d3bff46.mp4

**Comment by zekth** [comment]
This is more related to your code / dev env. How you're running it? Do you have a reproduceable example?
Here you can see the cold start benchmarks: https://github.com/fastify/benchmarks/blob/master/METRICS.md

**Reply by meotimdihia** [reply]
Thanks for the fast response. This is my example:
```node
server.register(require("./routes/common/redirect"))
server.register(require("./routes/common/languages_count"))
server.register(require("./routes/common/report_error"))
server.register(require("./routes/common/top"))
server.register(require("./routes/common/check_email"))

server.register(function(fastify, opts, done) {
  fastify.addHook("preHandler", fastify.auth([fastify.authenticated]))
  fastify.register(require("./routes/common/last_read"))
  fastify.register(require("./routes/reaction/get"))
  fastify.register(require("./routes/reaction/post"))
  fastify.register(require("./routes/follow/post"))
  fastify.register(require("./routes/user/follow/comic/ids"))
  fastify.register(require("./routes/comic/_id/post"))
  fastify.register(require("./routes/comic/post"))

  fastify.register(require("./routes/comic/revision_accept"))
  
  fastify.register(require("./routes/comic/_slug/add_chapter"))
  fastify.register(require("./routes/comic/_slug/rollback"))

  fastify.register(require("./routes/affiliate/post"))
  fastify.register(require("./routes/affiliate/_id/put"))
  fastify.register(require("./routes/affiliate/_id/delete"))
  fastify.register(require("./routes/comment/post"))
  

  done()
})


const ajv = new Ajv({
  removeAdditional: true,
  useDefaults: true,
  coerceTypes: "array",
  nullable: true
})
server.setValidatorCompiler(({ schema }) => {
  return ajv.compile(schema)
})

const start = async () => {
  try {
    await server.listen(3001, "0.0.0.0")
    server.swagger()
  } catch (err) {
    server.log.error(err)
    process.exit(1)
  }
}
start()
```

**Reply by meotimdihia** [reply]
My routes is usualy like this: 
```node
import { FastifyInstance } from "fastify"

import { FromSchema } from "json-schema-to-ts"

module.exports = async (fastify: FastifyInstance) => {
  const querystring = {
    type: "object",
    require: ["filter"],
    properties: {
      filter: { type: "string" }
    }
  } as const
  fastify.get<{ Querystring: FromSchema<typeof querystring> }>(
    "/test",
    {
      schema: {
        querystring: querystring,
        response: {
          200: {
            type: "array",
            additionalProperties: true
          }
        }
      }
    },
    async (req, reply) => {
      const filter = JSON.parse(req.query.filter)
```

**Comment by delvedor** [comment]
Hello! We would need to know more about your environment, what your you doing with Fastify and how are you using it.
My gut feeling is that TypeScript is taking a lot of time.
What happens if you compile the project with `tsc` (not `ts-node`) and then run it?

**Reply by meotimdihia** [reply]
https://user-images.githubusercontent.com/300961/142990812-d26414e5-d544-4b6e-bce5-40f30adfcabc.mp4

I have tried to run `tsc` first and have used `node` to start the server, it still takes 5s to start for me.

**Reply by meotimdihia** [reply]
I don't think my code has any strange things. I have used `Prisma` and I added examples in the above answer.

**Reply by delvedor** [reply]
You can use https://clinicjs.org/ to understand what's causing the slowdown.

**Comment by mcollina** [comment, marked as answer]
Thanks for reporting this. I think there are three problems:

1. `tsc` is taking significant time to compile the files. I _think_ it's related to `json-schema-to-ts`.
2. Loading a lot of JSON Schemas inside Ajv is expensive. This is a known problem and @zekth is working on it. Hopefully this will improve in Fastify v4 or a follow-up. (We have a proof-of-concept that significantly improve this).
3. There is a database connection somewhere taking some time
